### PR TITLE
[APPINT-1339] Improving Connector Framework

### PR DESCRIPTION
**Pull request by hhemanth**
- Made changed to Gemfile add new gems and remove some gems from production env
- Add puma configs
- Added default README
- Added circleci config

### CODE REVIEW

File: template/files/circleci_config.yml [hunk 1]
```diff
@@ -0,0 +1,69 @@
+# Ruby CircleCI 2.0 configuration file
+#
+# Check https://circleci.com/docs/2.0/language-ruby/ for more details
+#
+version: 2
+jobs:
+  build:
+    docker:
+      # specify the version you desire here
+      - image: circleci/ruby:2.3-node
```
Comment (adamaziz15) [inline template/files/circleci_config.yml]: Might want to replace the image with:
`image: circleci/ruby:2.3-stretch-node`

Not completely sure it's still an issue but there was a circleci cache issue we were having that required us to re-trigger some builds as they'd fail on the first run: https://support.circleci.com/hc/en-us/articles/360004937834-Resolving-Ruby-Jessie-to-Stretch-issue-cannot-open-shared-object-file-cache-problem

File: template/files/circleci_config.yml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        environment:
+          - RAILS_ENV: test
+          - REDIS_URL: "redis://localhost:6379/connector"
```
Comment (adamaziz15) [inline template/files/circleci_config.yml]: Do we need the `REDIS_URL` here?

File: template/files/circleci_config.yml (hunk 1 above)
Comment (hhemanth) [reply template/files/circleci_config.yml]: Will do so

File: template/files/circleci_config.yml (hunk 2 above)
Comment (hhemanth) [reply template/files/circleci_config.yml]: I took the config from marlin-banking-adapter.  shall I remove it?

File: template/files/README.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
+## Development Setup
+ ### Setup the project
```
Comment (ouranos) [inline template/files/README.md]: Add some spacing between sections

File: template/files/Gemfile.tt [hunk 4]
```diff
@@ -5,6 +5,7 @@ gem 'tzinfo-data', platforms: [:mingw, :mswin, :jruby]
 
 gem 'puma', require: false
 
+gem 'redis-rails'
```
Comment (ouranos) [inline template/files/Gemfile.tt]: Why did you add `redis-rails`?
Is it required by something else?

File: template/files/README.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+```
+ Update the configuration in  `config/application.yml`.
+ ### Start the server
+ ```
+foreman start
+```
+ ### Run the test suite
+ ```
+bin/rake
+```
```
Comment (ouranos) [inline template/files/README.md]: Could you add a placeholder section with `TODO` for the developer to explain how to get a developer API key from the external app. See https://github.com/maestrano/connector-salesforce#configuration or the ones you added for the other connectors.

See the way the original `README.rdoc` is asking the developer to add details how the application.

File: template/files/Gemfile.tt [hunk 6]
```diff
@@ -17,22 +17,24 @@ gem 'pundit'
 gem 'pundit-resources'
 
 group :production, :uat do
-  gem 'activerecord-jdbcmysql-adapter', platforms: :jruby
   # Lock mysql2 to an old version for now as there is a bug in Rails 4.1.x and 4.2.x
   # The bug prevents activerecord from using the most recent version of mysql2.
```
Comment (hhemanth) [inline template/files/Gemfile.tt]: This comment is no longer valid, I will remove it and verify again

File: template/files/Gemfile.tt [hunk 7]
```diff
@@ -17,22 +17,24 @@ gem 'pundit'
...
   # See: https://stackoverflow.com/questions/32457657/rails-4-gemloaderror-specified-mysql2-for-database-adapter-but-the-gem-i
-  gem 'mysql2', '~> 0.3.18', platforms: :ruby
+  gem 'mysql2', '~> 0.4.10', platforms: :ruby
```
Comment (hhemanth) [inline template/files/Gemfile.tt]: this is due to mysql version on macosx, will add a comment to that effect

File: maestrano-connector-rails.gemspec [hunk 8]
```diff
@@ -31,7 +31,7 @@ Gem::Specification.new do |s|
   s.add_runtime_dependency('bootstrap-sass')
   s.add_runtime_dependency('config')
   # v1.1 breaks the build. See: https://github.com/ruby-concurrency/concurrent-ruby/issues/768
-  s.add_runtime_dependency('concurrent-ruby', '~> 1.0.5')
+  s.add_runtime_dependency('concurrent-ruby', '1.0.5')
```
Comment (ouranos) [inline maestrano-connector-rails.gemspec]: `~> 1.0.5` shouldn't have let you upgrade to 1.1?
Is `1.1.1` fixing it?

File: template/files/README.md [hunk 9]
```diff
@@ -0,0 +1,29 @@
+## Development Setup
+
+ ### Setup the project
+ ```
+git clone <github url of the repo>
+
+bin/setup
+```
+Update the configuration in  `config/application.yml`.
```
Comment (ouranos) [inline template/files/README.md]: 2 spaces

File: template/files/README.md [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+ ### How to setup expernal app and get keys
```
Comment (ouranos) [inline template/files/README.md]: Typo

File: template/files/README.md [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+ - How to create security credentials, like oauth client-id and client-secret.
+ - Where to populate the keys/credentials in the app, this will probably be application-sample.yml.
+ - Create variables in application-sample.yml to hold the keys and the URL of the 3rd party app. 
+
+ ### Start the server
+ ```
+foreman start
+```
+
+
```
Comment (ouranos) [inline template/files/README.md]: Too many whitelines

File: template/files/setup [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+Dir.chdir APP_ROOT do
+  # This script is a starting point to setup your application.
+  # Add necessary setup steps to this file:
+  # Do not try to load configuration from the developer platform
+  ENV['SKIP_CONFIGURATION'] = 'true'
+  puts '== Installing dependencies =='
+  system 'gem install bundler --conservative'
+  system 'bundle check || bundle install'
+
+  # puts '\n== Copying sample files =='
```
Comment (ouranos) [inline template/files/setup]: This section cab be removed

File: template/files/circleci_config.yml (hunk 1 above)
Comment (hhemanth) [reply template/files/circleci_config.yml]: done

File: template/files/circleci_config.yml (hunk 2 above)
Comment (hhemanth) [reply template/files/circleci_config.yml]: removed

File: template/files/README.md (hunk 3 above)
Comment (hhemanth) [reply template/files/README.md]: done

File: template/files/Gemfile.tt (hunk 4 above)
Comment (hhemanth) [reply template/files/Gemfile.tt]: added it as it wa sin banking-adapter, removed

File: maestrano-connector-rails.gemspec (hunk 8 above)
Comment (hhemanth) [reply maestrano-connector-rails.gemspec]: ~> 1.0.5 upgraded to 1.1.3 . 1.1.1 does not solve it so I am locking it to 1.0.5 for now.

File: template/files/circleci_config.yml [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+      - restore_cache:
+          keys:
+          - v1-dependencies-{{ checksum "Gemfile.lock" }}
+          # fallback to using the latest cache if no exact match is found
+          - v1-dependencies-
+
+      - run:
+          name: install dependencies
+          command: |
+            bundle install --jobs=4 --retry=3 --path vendor/bundle
```
Comment (alexmaestrano) [inline template/files/circleci_config.yml]: 👍

File: template/files/setup [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+  # Do not try to load configuration from the developer platform
+  ENV['SKIP_CONFIGURATION'] = 'true'
+  puts '== Installing dependencies =='
+  system 'gem install bundler --conservative'
+  system 'bundle check || bundle install'
+
+  puts '\n== Copying sample files =='
+  system 'cp config/database.yml.sample config/database.yml' unless File.exist?('config/database.yml')
+
+  system 'cp config/application-sample.yml config/application.yml' unless File.exist?('config/application.yml')
```
Comment (alexmaestrano) [inline template/files/setup]: 👍

File: maestrano-connector-rails.gemspec (hunk 8 above)
Comment (ouranos) [reply maestrano-connector-rails.gemspec]: As discussed this shouldn't be needed. See if it's better with `v2.3.6`

File: maestrano-connector-rails.gemspec [hunk 15]
```diff
@@ -13,7 +13,7 @@ Gem::Specification.new do |s|
   s.homepage     = "https://github.com/maestrano/maestrano-connector-rails"
   s.license      = "MIT"
 
-  s.required_ruby_version = '>= 2.3.0'
+  s.required_ruby_version = '>= 2.3.6'
```
Comment (ouranos) [inline maestrano-connector-rails.gemspec]: Is this bump required?
This would be a breaking change

**Comment by ouranos** [comment]
@hhemanth LGTM. could you squash those commits before I merge? Thanks.

**Comment by hhemanth** [comment]
> @hhemanth LGTM. could you squash those commits before I merge? Thanks.

@ouranos , Its squashed , You can merge now..
